File: lib/board.cpp
/**
 * board.cpp
 * by Yiheng Quan
 */

#include <iostream>
#include <sstream>
#include <queue>
#include <unordered_map>
#include <set>
#include "board.h"

/// Constructors
PBoard::PBoard() {}
PBoard::PBoard(const Board &board, int row, int column, int minEraseCondition)
{
    this->board = board;
    this->row = row;
    this->column = column;
    this->minEraseCondition = minEraseCondition;
}

PBoard::~PBoard()
{
    for (auto &row : board)
    {
        row.clear();
    }
    board.clear();
}

/// Board related

ComboList PBoard::eraseComboAndMoveOrbs(int *moveCount)
{
    ComboList combo;

    bool hasCombo;
    do
    {
        // Remember to reset
        hasCombo = false;
        for (int i = 0; i < column; i++)
        {
            for (int j = 0; j < row; j++)
            {
                // Ignore empty orbs
                if (board[i][j] == pad::empty)
                    continue;

                // Start finding combos
                auto erased = eraseCombo(&combo, i, j);
                if (erased)
                {
                    // Just to prevent setting it to false again
                    hasCombo = true;
                }
            }
        }

        // Only move orbs down if there is at least one combo
        if (hasCombo)
        {
            moveOrbsDown();
            *moveCount += 1;
        }
    } while (hasCombo);

    return combo;
}

// From https://stackoverflow.com/a/20602159
struct PairHash
{
public:
    template <typename T, typename U>
    std::size_t operator()(const std::pair<T, U> &x) const
    {
        return std::hash<T>()(x.first) ^ std::hash<U>()(x.second);
    }
};

bool PBoard::eraseCombo(ComboList *list, int ox, int oy)
{
    using namespace std;
    queue<OrbLocation> toVisit;
    set<OrbLocation> inserted;
    toVisit.emplace(LOCATION(ox, oy));
    unordered_map<OrbLocation, bool, PairHash> visited;

    // Keep searching until all locations are visited
    while (toVisit.size() > 0)
    {
        auto currentLocation = toVisit.front();
        toVisit.pop();

        // Don't visit the same location twice
        if (visited[currentLocation])
            continue;

        int x = currentLocation.first, y = currentLocation.second;
        auto orb = board[x][y];
        // Look vertically
        int up = x, down = x;
        int upOrb = 0, downOrb = 0;
        while (--up >= 0)
        {
            // Break immediately if nothing matches, a deadend
            if (hasSameOrb(orb, up, y))
                upOrb++;
            else
                break;
        }
        while (++down < column)
        {
            if (hasSameOrb(orb, down, y))
                downOrb++;
            else
                break;
        }
        // We need to have at least 3 (eraseCondition) connected
        if (upOrb + downOrb + 1 >= minEraseCondition)
        {
            // Add them in connected orbs
            for (int i = x - upOrb; i < x + downOrb; i++)
            {
                auto l = LOCATION(i, y);
                // Don't add to connected orbs multiple times
                if (visited[l])
                    continue;
                inserted.insert(l);
                toVisit.push(l);
            }
        }

        // Look horizontally
        int left = y, right = y;
        int leftOrb = 0, rightOrb = 0;
        while (--left >= 0)
        {
            if (hasSameOrb(orb, x, left))
                leftOrb++;
            else
                break;
        }
        while (++right < row)
        {
            if (hasSameOrb(orb, x, right))
                rightOrb++;
            else
                break;
        }
        // Same as above
        if (leftOrb + rightOrb + 1 >= minEraseCondition)
        {
            // Add them in connected orbs
            for (int i = y - leftOrb; i < y + rightOrb; i++)
            {
                auto l = LOCATION(x, i);
                // Don't add to connected orbs multiple times
                if (visited[l])
                    continue;
                inserted.insert(l);
                toVisit.push(l);
            }
        }

        // Finally set currentLocation to be visited
        visited[currentLocation] = true;
    }

    // Erase all connected orbs
    bool hasCombo = inserted.size() >= minEraseCondition;
    if (hasCombo)
    {
        Combo combo;
        combo.reserve(inserted.size());
        for (const auto &l : inserted)
        {
            board[l.first][l.second] = pad::empty;
            combo.push_back(l);
        }

        list->push_back(combo);
    }
    return hasCombo;
}

int PBoard::rateBoard()
{
    int score = 0;

    // total combo
    int combo = 0;
    // If you move more, you get more combo in real life but sometimes not
    int moveCount = 0;
    // int newCombo = 0;
    // do
    // {
    //     newCombo = eraseOrbs();
    //     if (newCombo > 0)
    //     {
    //         combo += newCombo;
    //         // Even if it has at least one new combo
    //         moveOrbsDown();
    //         moveCount++;
    //     }
    // } while (newCombo > 0);
    auto list = eraseComboAndMoveOrbs(&moveCount);
    combo = list.size();

    // See if the remaining orbs are close to each other
    int orbAround = 0;
    int orbNearby = 0;
    for (int i = 0; i < column; i++)
    {
        for (int j = 0; j < row; j++)
        {
            auto curr = board[i][j];
            if (curr == pad::empty)
                continue;
            // Check if there are same orbs around
            for (int a = -1; a <= 1; a++)
            {
                for (int b = -1; b <= 1; b++)
                {
                    // This is the current orb
                    if (a == 0 && b == 0)
                        continue;
                    if (hasSameOrb(curr, i + a, j + b))
                    {
                        orbAround++;
                        if ((a == 0 && ((b == 1) || (b == -1))) ||
                            (b == 0 && ((a == 1) || (a == -1))))
                        {
                            // This means that it is a line
                            orbNearby += 1;
                            orbAround -= 1;
                        }
                    }
                }
            }
        }
    }

    score += pad::ORB_AROUND_SCORE * orbAround;
    score += pad::ORB_NEARBY_SCORE * orbNearby;

    score += pad::ONE_COMBO_SCORE * combo;
    score += pad::CASCADE_SCORE * moveCount;

    // Try out amen puzzle
    // int orbLeft = 0;
    // for (int i = 0; i < column; i++)
    // {
    //     for (int j = 0; j < row; j++)
    //     {
    //         auto curr = board[i][j];
    //         if (curr != pad::empty)
    //             orbLeft++;
    //     }
    // }

    // int comboMoreOrLess = abs(combo - 7);
    // int orbLeftMore = orbLeft - 3;
    // if (comboMoreOrLess == 0 && orbLeftMore <= 0)
    // {
    //     score += 7000;
    // }

    // score -= orbLeftMore * pad::ONE_COMBO_SCORE;
    // score -= comboMoreOrLess * pad::CASCADE_SCORE;

    if (printMoreMessages)
        std::cout << "That was " << combo << " combo\n";
    return score;
}

void PBoard::moveOrbsDown()
{
    for (int j = 0; j < row; j++)
    {
        std::vector<pad::orbs> orbs;
        orbs.reserve(column);
        int emptyCount = 0;
        // Start checking from the bottom most column
        for (int i = column - 1; i >= 0; i--)
        {
            auto orb = board[i][j];
            if (pad::empty != orb)
            {
                orbs.push_back(orb);
            }
            else
            {
                emptyCount++;
            }
        }

        // Must be less than column (it means that this column is all empty)
        if (emptyCount > 0 && emptyCount < column)
        {
            // Fill the saved orbs
            int k = 0, s = (int)orbs.size();
            for (int i = column - 1; i >= 0; --i, ++k)
            {
                if (k >= s)
                {
                    board[i][j] = pad::empty;
                }
                else
                {
                    // If column is 5, i starts from 4 so the index of orb is 5 - 1 - 4 = 0
                    auto orb = orbs[k];
                    board[i][j] = orb;
                }
            }
        }
    }
}

int PBoard::eraseOrbs()
{
    int combo = 0;

    // Collect all orbs that can be erased
    OrbSet orbs;

    for (int i = column - 1; i >= 0; i--)
    {
        for (int j = 0; j < row; j++)
        {
            auto curr = board[i][j];
            // Ignore empty orbs
            if (curr == pad::empty)
                continue;

            // vertical and horizontal orbs
            auto vhOrbs = findSameOrbsAround(i, j);
            // Here we need to loop throufh vhOrbs and check all orbs to see if there are orbs that can be erased
            auto it = vhOrbs.begin();
            while (it != vhOrbs.end())
            {
                // nextOrb is a pointer to the next pair
                auto nextOrb = nextSameOrbAround(&vhOrbs, *it);
                // Only search if there are new orbs, if nothing has been found (-1, -1) is returned
                if (nextOrb.first > -1)
                {
                    auto newOrbs = findSameOrbsAround(nextOrb);
                    vhOrbs.insert(newOrbs.begin(), newOrbs.end());
                    // Must check if there are new orbs or it will be an infinite loop
                    if (newOrbs.size() > 0)
                    {
                        // There are at least some orbs around it reset and continue
                        // TODO: ideally you want to start from new orbs but how to achieve it?
                        it = vhOrbs.begin();
                        continue;
                    }
                }
                it++;
            }

            // There should be orbs inside, check if the size is more than minEraseCondition (it was causing some issues)
            if ((int)vhOrbs.size() >= minEraseCondition)
            {
                for (auto const &xy : vhOrbs)
                {
                    ORB(board, xy) = pad::empty;
                }
                combo++;
            }
        }
    }

    return combo;
}

OrbSet PBoard::findSameOrbsAround(int x, int y)
{
    auto curr = board[x][y];

    // Check vertically
    OrbSet vOrbs;
    // Add this orb first or another infinite loop
    vOrbs.insert(LOCATION(x, y));
    int up = x, down = x;
    int upOrb = 1, downOrb = 1;
    while (--up >= 0)
    {
        if (board[up][y] == curr)
        {
            vOrbs.insert(LOCATION(up, y));
            upOrb++;
        }
        else
        {
            // Break immediately if nothing matches
            break;
        }
    }
    while (++down < column)
    {
        if (board[down][y] == curr)
        {
            vOrbs.insert(LOCATION(down, y));
            downOrb++;
        }
        else
        {
            break;
        }
    }
    // Now, we only need to check if there are at least 3 (4 or 5) orbs vertically
    // Less than the condition, -1 to remove the duplicate (which is the current orb)
    if (upOrb + downOrb - 1 < minEraseCondition)
        vOrbs.clear();

    // Check horizontally
    OrbSet hOrbs;
    // Add it again just in case it was cleared
    hOrbs.insert(LOCATION(x, y));
    int left = y, right = y;
    int leftOrb = 1, rightOrb = 1;
    while (--left >= 0)
    {
        if (board[x][left] == curr)
        {
            hOrbs.insert(LOCATION(x, left));
            leftOrb++;
        }
        else
        {
            break;
        }
    }
    while (++right < row)
    {
        if (board[x][right] == curr)
        {
            hOrbs.insert(LOCATION(x, right));
            rightOrb++;
        }
        else
        {
            break;
        }
    }
    // Same as above
    if (leftOrb + rightOrb - 1 < minEraseCondition)
        hOrbs.clear();

    // Merge vertical and horizontal all to vertical
    vOrbs.insert(hOrbs.begin(), hOrbs.end());
    return vOrbs;
}

OrbLocation PBoard::nextSameOrbAround(OrbSet *vhOrbs, int x, int y)
{
    auto orb = board[x][y];

    // Find up, down, left and right
    if (hasSameOrb(orb, x - 1, y))
    {
        OrbLocation pair = LOCATION(x - 1, y);
        if (vhOrbs->count(pair) == 0)
            return pair;
    }
    if (hasSameOrb(orb, x + 1, y))
    {
        OrbLocation pair = LOCATION(x + 1, y);
        if (vhOrbs->count(pair) == 0)
            return pair;
    }
    if (hasSameOrb(orb, x, y - 1))
    {
        OrbLocation pair = LOCATION(x, y - 1);
        if (vhOrbs->count(pair) == 0)
            return pair;
    }
    if (hasSameOrb(orb, x, y + 1))
    {
        OrbLocation pair = LOCATION(x, y + 1);
        if (vhOrbs->count(pair) == 0)
            return pair;
    }

    // Return a default value
    return LOCATION(-1, -1);
}

void PBoard::printBoard()
{
    if (isEmptyFile())
    {
        std::cout << "- empty -\n";
        return;
    }

    // Print everything out nicely
    std::cout << std::endl;
    std::cout << row << " x " << column << std::endl;
    for (auto const &row : board)
    {
        for (auto const &orb : row)
        {
            std::cout << pad::ORB_NAMES[orb] << "\t";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

void PBoard::printBoardForSimulation()
{
    std::stringstream ss;
    for (auto const &row : board)
    {
        for (auto const &orb : row)
        {
            std::cout << pad::ORB_SIMULATION_NAMES[orb];
            ss << (int)(orb - 1);
        }
    }
    std::cout << std::endl;
    std::cout << ss.str() << std::endl;
}

void PBoard::printBoardInfo()
{
    if (isEmptyFile())
    {
        std::cout << "no info\n";
        return;
    }

    // Collect orb info
    int *counter = collectOrbCount();

    // Print out some board info
    for (int i = 1; i < pad::ORB_COUNT; i++)
    {
        int count = counter[i];
        if (count == 0)
            continue;
        // It is just like fire x 5 wood x 6
        std::cout << count << " x " << pad::ORB_NAMES[i];
        if (i != pad::ORB_COUNT - 1)
            std::cout << " | ";
    }
    std::cout << std::endl;

    std::cout << "Board max combo: " << getBoardMaxCombo() << std::endl;
    std::cout << "Current max combo: " << getMaxCombo(counter) << std::endl;

    delete[] counter;
}

/// Utils

int PBoard::getMaxCombo(int *counter)
{
    if (isEmptyFile())
        return 0;

    int comboCounter = 0;

    int moreComboCount;
    do
    {
        // Reset this flag everytime
        moreComboCount = 0;
        // This tracks how many orbs are left
        int orbLeft = 0;
        int maxOrbCounter = 0;

        for (int i = 1; i < pad::ORB_COUNT; i++)
        {
            // Keep -3 or other minEraseCondition (4, 5) until all orbs are less than 2
            int curr = counter[i];
            if (curr >= minEraseCondition)
            {
                moreComboCount += 1;
                comboCounter++;
                counter[i] -= minEraseCondition;
                if (curr > maxOrbCounter)
                    maxOrbCounter = curr;
            }
            else
            {
                orbLeft += curr;
            }
        }

        // Only one colour can do combo now
        if (moreComboCount == 1)
        {
            // Orbs left are in different colour but they can still seperate other colours
            int maxComboPossible = orbLeft / minEraseCondition;
            int maxCombo = maxOrbCounter / minEraseCondition;
            // Make sure there are enough orbs
            comboCounter += maxCombo > maxComboPossible ? maxComboPossible : maxCombo;

            // No orbs left means one colour, -1 here because we count one more but doesn't apply to a single colour board
            if (orbLeft > 0)
                comboCounter--;
            break;
        }
    } while (moreComboCount > 0);

    return comboCounter;
}

File: python/board.py
"""
board.py
- data class for board to store the board
"""

from constant import ORB_SIMULATION_NAMES, Orb
from copy import copy, deepcopy
from random import randint


class Combo:
    loc = None
    orb = None

    def __init__(self, loc: tuple, orb: Orb):
        self.loc = loc
        self.orb = orb


class Board:
    row = 0
    column = 0
    board = []
    board_hash = ""

    min_erase = 3
    depth = 30
    width = 1000

    step = 0
    score = 0
    children = []
    loc = None

    parent = None
    prev_score = 0
    prev_loc = None

    def __init__(self, board_str, min_erase, depth, width):
        length = len(board_str)
        if length == 20:
            self.column, self.row = 4, 5
        elif length == 30:
            self.column, self.row = 5, 6
        elif length == 42:
            self.column, self.row = 6, 7
        else:
            raise AssertionError("Board length must be 20, 30 or 42")

        for x in range(self.column):
            new_row = []
            for y in range(self.row):
                orb_char = board_str[x * self.row + y]
                new_orb = Orb(ORB_SIMULATION_NAMES.index(orb_char))
                new_row.append(new_orb)
            self.board.append(new_row)

        self.board_hash = board_str
        self.min_erase = min_erase
        self.depth = depth
        self.width = width

    def __lt__(self, other):
        if self.score == other.score:
            return self.step > other.step
        return self.score < other.score

    def swap(self, from_loc, to_loc):
        """
        copy current board and swap 2 orb locations
        """

        board_copy = self.duplicate()
        new_board = board_copy.board
        board_copy.children = []

        # swap two values, so simple in python
        x, y = from_loc
        nx, ny = to_loc
        new_board[x][y], new_board[nx][ny] = new_board[nx][ny], new_board[x][y]

        board_copy.step += 1
        board_copy.loc = to_loc

        board_copy.parent = self
        board_copy.prev_score = self.score
        board_copy.prev_loc = self.loc

        board_copy.calc_board_hash(force=True)
        board_copy.calc_score()

        return board_copy

    def get_children(self, diagonal=False):
        if self.step == self.depth:
            return

        for x in range(-1, 2):
            for y in range(-1, 2):
                if x == y == 0:
                    continue

                is_diagonal = abs(x) * abs(y) > 0
                # continue if it is diagonal but not in diagonal mode
                if not diagonal and is_diagonal:
                    continue

                nx, ny = self.loc
                nx += x
                ny += y
                if (nx, ny) == self.prev_loc:
                    continue
                if self.is_valid_loc((nx, ny)):
                    new_board = self.swap(self.loc, (nx, ny))
                    self.children.append(new_board)

    def is_valid_loc(self, loc):
        # check x and y are in bound
        x, y = loc
        return 0 <= x <= self.column - 1 and 0 <= y <= self.row - 1

    def calc_board_hash(self, force=False):
        # only calculate if current hash is empty
        if len(self.board_hash) == 0 or force:
            self.board_hash = ""
            for x in self.board:
                for y in x:
                    self.board_hash += ORB_SIMULATION_NAMES[y.value]

    def calc_score(self):
        # copy the board to erase combo
        temp = deepcopy(self.board)
        combo_list = []
        combo = []
        move_count = 0

        more_combo = True
        while more_combo:
            more_combo = False

            for x in range(self.column - 1, -1, -1):
                for y in range(self.row):
                    curr_orb = temp[x][y]
                    if curr_orb == Orb.EMPTY:
                        continue

                    self._flood_fill(temp, combo, x, y, curr_orb, -1)
                    if len(combo) >= self.min_erase:
                        more_combo = True
                        combo_list.append(combo)
                        combo = []

            if more_combo:
                move_count += 1
                more_combo = self._move_orbs_down(temp)
        self.score = len(combo_list) * 1000 + move_count * 50

    def _flood_fill(self, board, combo, x, y, orb, direction):
        if not self.is_valid_loc((x, y)):
            return

        curr_orb = board[x][y]
        if curr_orb != orb:
            return

        count = 0
        direction_list = [0, 0, 0, 0]

        for d in range(4):
            loop = self.row
            if d > 1:
                loop = self.column

            for i in range(loop):
                cx, cy = x, y
                if d == 0:
                    cy += i
                elif d == 1:
                    cy -= i
                elif d == 2:
                    cx += i
                elif d == 3:
                    cx -= i

                if not self.is_valid_loc((cx, cy)):
                    break
                if board[cx][cy] != orb:
                    break

                direction_list[d] += 1
                count += 1

        # added the same orb 4 times so need to -3
        count -= 3
        if count >= self.min_erase:
            min_connection = 3 if self.min_erase >= 3 else self.min_erase

            for d in range(4):
                curr_count = direction_list[d]
                if curr_count < min_connection:
                    continue

                for i in range(curr_count):
                    cx, cy = x, y
                    if d == 0:
                        cy += i
                    elif d == 1:
                        cy -= i
                    elif d == 2:
                        cx += i
                    elif d == 3:
                        cx -= i

                    has_orb_around = False
                    if d < 2:
                        has_orb_around = self.has_same_orb(
                            board, orb, cx + 1, cy) and self.has_same_orb(board, orb, cx - 1, cy)
                    else:
                        has_orb_around = self.has_same_orb(
                            board, orb, cx, cy + 1) and self.has_same_orb(board, orb, cx, cy - 1)

                    if not has_orb_around:
                        board[cx][cy] = Orb.EMPTY
                        combo.append(Combo((cx, cy), orb))

                for i in range(curr_count):
                    cx, cy = x, y
                    if d == 0:
                        cy += i
                    elif d == 1:
                        cy -= i
                    elif d == 2:
                        cx += i
                    elif d == 3:
                        cx -= i

                    self._flood_fill(board, combo, cx, cy + 1, orb, 0)
                    self._flood_fill(board, combo, cx, cy - 1, orb, 1)
                    self._flood_fill(board, combo, cx + 1, cy, orb, 2)
                    self._flood_fill(board, combo, cx - 1, cy, orb, 3)

    def has_same_orb(self, board, orb, x, y):
        if self.is_valid_loc((x, y)):
            return board[x][y] == orb
        return False

    def _move_orbs_down(self, board):
        changed = False
        for y in range(self.row):
            orbs = []
            empty = 0

            # from bottom up
            for x in range(self.column - 1, -1, -1):
                curr_orb = board[x][y]
                if curr_orb == Orb.EMPTY:
                    empty += 1
                else:
                    orbs.append(curr_orb)

            # at least one orb is removed but do nothing if all are removed
            if 0 < empty < self.column:
                changed = True
                s = len(orbs)
                k = 0
                for x in range(self.column - 1, -1, -1):
                    if k >= s:
                        board[x][y] = Orb.EMPTY
                    else:
                        board[x][y] = orbs[k]
                    k += 1
        return changed

    def info(self):
        self.print_board()
        print("{} x {} | Step - {} | {} -> {} | Score - {} ({}) | Children - {}"
              .format(self.row, self.column, self.step, self.prev_loc, self.loc, self.score, self.prev_score, len(self.children)))

    def duplicate(self):
        new_copy = copy(self)
        new_copy.board = deepcopy(self.board)
        return new_copy

    def get_target_score(self):
        # an estimation of the score we want to reach
        return 8000  # 8 combo

    def print_board(self, number=False, offset=1):
        if not number:
            print(self.board_hash)
        else:
            for x in self.board:
                for y in x:
                    print(y.value - offset, end="")
            print()
